File: python/src/multifrost/core/metrics.py
"""
Metrics collection for observability.

Tracks request latency, error rates, queue depth, and circuit breaker events.
"""

import time
from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional
from collections import deque
import threading


@dataclass
class RequestMetrics:
    """Metrics for a single completed request."""

    request_id: str
    function: str
    namespace: str
    success: bool
    latency_ms: float
    error: Optional[str] = None
    timestamp: float = field(default_factory=time.time)


@dataclass
class MetricsSnapshot:
    """Point-in-time snapshot of all metrics."""

    # Counters
    requests_total: int = 0
    requests_success: int = 0
    requests_failed: int = 0

    # Latency (milliseconds)
    latency_avg_ms: float = 0.0
    latency_p50_ms: float = 0.0
    latency_p95_ms: float = 0.0
    latency_p99_ms: float = 0.0
    latency_min_ms: float = 0.0
    latency_max_ms: float = 0.0

    # Queue
    queue_depth: int = 0
    queue_max_depth: int = 0

    # Circuit breaker
    circuit_breaker_trips: int = 0
    circuit_breaker_state: str = "closed"  # closed, open, half-open

    # Time window
    window_seconds: float = 0.0

    # Timestamp
    timestamp: float = field(default_factory=time.time)


class Metrics:
    """
    Thread-safe metrics collector for ParentWorker.

    Usage:
        metrics = Metrics()

        # Record a request
        with metrics.track_request("my_func", "default") as tracker:
            result = await call_remote()
            tracker.success()

        # Or manual tracking
        start = metrics.start_request("my_func", "default")
        # ... do work ...
        metrics.end_request(start, success=True)

        # Get snapshot
        snapshot = metrics.snapshot()
        print(f"Avg latency: {snapshot.latency_avg_ms}ms")
    """

    def __init__(
        self,
        max_latency_samples: int = 1000,
        window_seconds: float = 60.0,
    ):
        self.max_latency_samples = max_latency_samples
        self.window_seconds = window_seconds

        # Counters (thread-safe via locks)
        self._lock = threading.Lock()
        self._requests_total = 0
        self._requests_success = 0
        self._requests_failed = 0
        self._circuit_breaker_trips = 0
        self._circuit_breaker_state = "closed"

        # Queue tracking
        self._queue_depth = 0
        self._queue_max_depth = 0

        # Latency samples (circular buffer)
        self._latencies: deque = deque(maxlen=max_latency_samples)

        # Request tracking (for in-flight requests)
        self._inflight: Dict[str, RequestMetrics] = {}

    def start_request(
        self,
        request_id: str,
        function: str,
        namespace: str = "default",
    ) -> float:
        """
        Start tracking a request.

        Returns start timestamp for later end_request() call.
        """
        with self._lock:
            self._requests_total += 1
            self._queue_depth += 1
            self._queue_max_depth = max(self._queue_max_depth, self._queue_depth)

        return time.perf_counter()

    def end_request(
        self,
        start_time: float,
        request_id: str,
        success: bool = True,
        error: Optional[str] = None,
    ) -> float:
        """
        End tracking a request.

        Returns latency in milliseconds.
        """
        latency_ms = (time.perf_counter() - start_time) * 1000

        with self._lock:
            self._queue_depth -= 1

            if success:
                self._requests_success += 1
            else:
                self._requests_failed += 1

            # Store latency sample
            self._latencies.append(latency_ms)

        return latency_ms

    def record_circuit_breaker_trip(self):
        """Record a circuit breaker trip event."""
        with self._lock:
            self._circuit_breaker_trips += 1
            self._circuit_breaker_state = "open"

    def record_circuit_breaker_reset(self):
        """Record circuit breaker reset (closed)."""
        with self._lock:
            self._circuit_breaker_state = "closed"

    def record_circuit_breaker_half_open(self):
        """Record circuit breaker entering half-open state."""
        with self._lock:
            self._circuit_breaker_state = "half-open"

    def snapshot(self) -> MetricsSnapshot:
        """Get a point-in-time snapshot of all metrics."""
        with self._lock:
            latencies = list(self._latencies)

            # Calculate percentiles
            if latencies:
                sorted_latencies = sorted(latencies)
                n = len(sorted_latencies)
                p50_idx = int(n * 0.50)
                p95_idx = int(n * 0.95)
                p99_idx = int(n * 0.99)

                latency_avg = sum(latencies) / n
                latency_p50 = sorted_latencies[min(p50_idx, n - 1)]
                latency_p95 = sorted_latencies[min(p95_idx, n - 1)]
                latency_p99 = sorted_latencies[min(p99_idx, n - 1)]
                latency_min = sorted_latencies[0]
                latency_max = sorted_latencies[-1]
            else:
                latency_avg = latency_p50 = latency_p95 = latency_p99 = 0.0
                latency_min = latency_max = 0.0

            return MetricsSnapshot(
                requests_total=self._requests_total,
                requests_success=self._requests_success,
                requests_failed=self._requests_failed,
                latency_avg_ms=latency_avg,
                latency_p50_ms=latency_p50,
                latency_p95_ms=latency_p95,
                latency_p99_ms=latency_p99,
                latency_min_ms=latency_min,
                latency_max_ms=latency_max,
                queue_depth=self._queue_depth,
                queue_max_depth=self._queue_max_depth,
                circuit_breaker_trips=self._circuit_breaker_trips,
                circuit_breaker_state=self._circuit_breaker_state,
                window_seconds=self.window_seconds,
            )

    def reset(self):
        """Reset all metrics."""
        with self._lock:
            self._requests_total = 0
            self._requests_success = 0
            self._requests_failed = 0
            self._circuit_breaker_trips = 0
            self._circuit_breaker_state = "closed"
            self._queue_depth = 0
            self._queue_max_depth = 0
            self._latencies.clear()
            self._inflight.clear()

    def to_dict(self) -> Dict[str, Any]:
        """Get metrics as a dictionary (for logging/serialization)."""
        snapshot = self.snapshot()
        return {
            "requests": {
                "total": snapshot.requests_total,
                "success": snapshot.requests_success,
                "failed": snapshot.requests_failed,
                "error_rate": (
                    snapshot.requests_failed / snapshot.requests_total
                    if snapshot.requests_total > 0
                    else 0.0
                ),
            },
            "latency_ms": {
                "avg": round(snapshot.latency_avg_ms, 2),
                "p50": round(snapshot.latency_p50_ms, 2),
                "p95": round(snapshot.latency_p95_ms, 2),
                "p99": round(snapshot.latency_p99_ms, 2),
                "min": round(snapshot.latency_min_ms, 2),
                "max": round(snapshot.latency_max_ms, 2),
            },
            "queue": {
                "depth": snapshot.queue_depth,
                "max_depth": snapshot.queue_max_depth,
            },
            "circuit_breaker": {
                "trips": snapshot.circuit_breaker_trips,
                "state": snapshot.circuit_breaker_state,
            },
        }

File: golang/metrics.go
package multifrost

import (
	"sort"
	"sync"
	"time"
)

// MetricsSnapshot represents a point-in-time snapshot of all metrics
type MetricsSnapshot struct {
	// Counters
	RequestsTotal   int `json:"requests_total"`
	RequestsSuccess int `json:"requests_success"`
	RequestsFailed  int `json:"requests_failed"`

	// Latency (milliseconds)
	LatencyAvgMs float64 `json:"latency_avg_ms"`
	LatencyP50Ms float64 `json:"latency_p50_ms"`
	LatencyP95Ms float64 `json:"latency_p95_ms"`
	LatencyP99Ms float64 `json:"latency_p99_ms"`
	LatencyMinMs float64 `json:"latency_min_ms"`
	LatencyMaxMs float64 `json:"latency_max_ms"`

	// Queue
	QueueDepth    int `json:"queue_depth"`
	QueueMaxDepth int `json:"queue_max_depth"`

	// Circuit breaker
	CircuitBreakerTrips int    `json:"circuit_breaker_trips"`
	CircuitBreakerState string `json:"circuit_breaker_state"`

	// Heartbeat
	HeartbeatRttAvgMs  float64 `json:"heartbeat_rtt_avg_ms"`
	HeartbeatRttLastMs float64 `json:"heartbeat_rtt_last_ms"`
	HeartbeatMisses    int     `json:"heartbeat_misses"`

	// Time window
	WindowSeconds float64 `json:"window_seconds"`

	// Timestamp
	Timestamp time.Time `json:"timestamp"`
}

// Metrics is a thread-safe metrics collector for ParentWorker
type Metrics struct {
	mu sync.RWMutex

	// Configuration
	maxLatencySamples int
	windowSeconds     float64

	// Counters
	requestsTotal       int
	requestsSuccess     int
	requestsFailed      int
	circuitBreakerTrips int
	circuitBreakerState string

	// Queue tracking
	queueDepth    int
	queueMaxDepth int

	// Latency samples (circular buffer via slice)
	latencies []float64

	// Heartbeat RTT samples
	heartbeatRtts   []float64
	heartbeatMisses int
}

// NewMetrics creates a new Metrics instance
func NewMetrics(maxLatencySamples int, windowSeconds float64) *Metrics {
	if maxLatencySamples <= 0 {
		maxLatencySamples = 1000
	}
	if windowSeconds <= 0 {
		windowSeconds = 60.0
	}

	return &Metrics{
		maxLatencySamples:   maxLatencySamples,
		windowSeconds:       windowSeconds,
		circuitBreakerState: "closed",
		latencies:           make([]float64, 0, maxLatencySamples),
		heartbeatRtts:       make([]float64, 0, 100),
	}
}

// StartRequest starts tracking a request
// Returns start timestamp for later EndRequest() call
func (m *Metrics) StartRequest(requestID, function, namespace string) time.Time {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.requestsTotal++
	m.queueDepth++
	if m.queueDepth > m.queueMaxDepth {
		m.queueMaxDepth = m.queueDepth
	}

	return time.Now()
}

// EndRequest ends tracking a request
// Returns latency in milliseconds
func (m *Metrics) EndRequest(startTime time.Time, requestID string, success bool, errorMsg string) float64 {
	latencyMs := float64(time.Since(startTime).Milliseconds())

	m.mu.Lock()
	defer m.mu.Unlock()

	m.queueDepth--

	if success {
		m.requestsSuccess++
	} else {
		m.requestsFailed++
	}

	// Store latency sample (circular buffer)
	if len(m.latencies) >= m.maxLatencySamples {
		// Remove oldest (shift left)
		m.latencies = m.latencies[1:]
	}
	m.latencies = append(m.latencies, latencyMs)

	return latencyMs
}

// RecordCircuitBreakerTrip records a circuit breaker trip event
func (m *Metrics) RecordCircuitBreakerTrip() {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.circuitBreakerTrips++
	m.circuitBreakerState = "open"
}

// RecordCircuitBreakerReset records circuit breaker reset (closed)
func (m *Metrics) RecordCircuitBreakerReset() {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.circuitBreakerState = "closed"
}

// RecordCircuitBreakerHalfOpen records circuit breaker entering half-open state
func (m *Metrics) RecordCircuitBreakerHalfOpen() {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.circuitBreakerState = "half-open"
}

// RecordHeartbeatRtt records a heartbeat round-trip time
func (m *Metrics) RecordHeartbeatRtt(rttMs float64) {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Keep last 100 samples
	if len(m.heartbeatRtts) >= 100 {
		m.heartbeatRtts = m.heartbeatRtts[1:]
	}
	m.heartbeatRtts = append(m.heartbeatRtts, rttMs)
}

// RecordHeartbeatMiss records a missed heartbeat (timeout)
func (m *Metrics) RecordHeartbeatMiss() {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.heartbeatMisses++
}

// Snapshot returns a point-in-time snapshot of all metrics
func (m *Metrics) Snapshot() MetricsSnapshot {
	m.mu.RLock()
	defer m.mu.RUnlock()

	snapshot := MetricsSnapshot{
		RequestsTotal:       m.requestsTotal,
		RequestsSuccess:     m.requestsSuccess,
		RequestsFailed:      m.requestsFailed,
		QueueDepth:          m.queueDepth,
		QueueMaxDepth:       m.queueMaxDepth,
		CircuitBreakerTrips: m.circuitBreakerTrips,
		CircuitBreakerState: m.circuitBreakerState,
		HeartbeatMisses:     m.heartbeatMisses,
		WindowSeconds:       m.windowSeconds,
		Timestamp:           time.Now(),
	}

	// Calculate latency percentiles
	if len(m.latencies) > 0 {
		latencies := make([]float64, len(m.latencies))
		copy(latencies, m.latencies)
		sort.Float64s(latencies)

		n := len(latencies)
		snapshot.LatencyMinMs = latencies[0]
		snapshot.LatencyMaxMs = latencies[n-1]

		// Calculate average
		sum := 0.0
		for _, v := range latencies {
			sum += v
		}
		snapshot.LatencyAvgMs = sum / float64(n)

		// Calculate percentiles
		snapshot.LatencyP50Ms = latencies[n*50/100]
		snapshot.LatencyP95Ms = latencies[n*95/100]
		snapshot.LatencyP99Ms = latencies[n*99/100]
	}

	// Calculate heartbeat RTT average
	if len(m.heartbeatRtts) > 0 {
		sum := 0.0
		for _, v := range m.heartbeatRtts {
			sum += v
		}
		snapshot.HeartbeatRttAvgMs = sum / float64(len(m.heartbeatRtts))
		snapshot.HeartbeatRttLastMs = m.heartbeatRtts[len(m.heartbeatRtts)-1]
	}

	return snapshot
}

// Reset resets all metrics
func (m *Metrics) Reset() {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.requestsTotal = 0
	m.requestsSuccess = 0
	m.requestsFailed = 0
	m.circuitBreakerTrips = 0
	m.circuitBreakerState = "closed"
	m.queueDepth = 0
	m.queueMaxDepth = 0
	m.latencies = make([]float64, 0, m.maxLatencySamples)
	m.heartbeatRtts = make([]float64, 0, 100)
	m.heartbeatMisses = 0
}
